saliency_metrics/metrics/sensitivity_n/sensn_reinference.py

```diff
@@ -0,0 +1,64 @@
+from typing import Dict
+
+import numpy as np
+import torch
+from sensn_perturbation import SensitivityNPerturbation
+from sensn_results import SensitivityNResult
+
+from saliency_metrics.metrics.build_metric import ReInferenceMetric
+from saliency_metrics.metrics.serializable_result import SerializableResult
+from saliency_metrics.models import freeze_module
+from saliency_metrics.models.build_classifier import build_classifier
+
+
+class SensitivityN(ReInferenceMetric):
+    def __init__(
+        self, classifier_cfg: Dict, log_n_max: int, log_n_ticks: float, summarized: bool = True, num_masks: int = 100
+    ) -> None:
+        self.classifier = build_classifier(classifier_cfg)
+        # freeze the model and turn eval mode on
+        freeze_module(self.classifier)
+
+        self._result: SerializableResult = SensitivityNResult(summarized=summarized)
+        self._num_masks = num_masks
+        n_list = np.logspace(0, log_n_max, int(log_n_max / log_n_ticks), base=10.0, dtype=int)
+
+        # to eliminate the duplicated elements caused by rounding
+        self._n_list = np.unique(n_list)
+        self._current_n_ind = 0
+        self._perturbation = SensitivityNPerturbation(self._n_list[self._current_n_ind], num_masks=self._num_masks)
+
+    @property
+    def num_ns(self):
+        return len(self._n_list)
+
+    @property
+    def current_n(self):
+        return self._n_list[self._current_n_ind]
+
+    def increment_n(self) -> None:
+        self._current_n_ind += 1
+        self._perturbation = SensitivityNPerturbation(self._n_list[self._current_n_ind], self._num_masks)
+
+    def evaluate(self, img: torch.Tensor, smap: torch.Tensor, target: int) -> Dict:
+        batched_sample, sum_attributions = self._perturbation.perturb(img, smap)
+
+        score_diffs = []
+        out = []
+        with torch.no_grad():
+            for batch in batched_sample:
+                scores = torch.softmax(self.classifier(batch.unsqueeze(0)), -1)[:, target]
+                out.append(scores.item())
+
+        for i, _ in enumerate(out[0:-1]):
```

Comment:
Too low efficient for-loop. If you feed the perturbed images along with the original image in a batch to the classifier. The score difference is then `scores[:-1] - scores[-1]`, where `scores` is the Tensor of the batched output. Note that we know the last component is associated with the original image.